opr/grader.go

```diff
@@ -104,23 +183,416 @@ func (g *Grader) Run(config *config.Config, monitor *common.Monitor) {
+		c := 0
+		// We have eblocks to sync!
+		// NextEBlock() will return the next eblock in the chain that we still need to sync
+		// it's entries. So we can walk, NextEblock -> NextEblock -> ... to sync the whole chain.
+		for block := g.OPRChain.NextEBlock(); block != nil; block = g.OPRChain.NextEBlock() {
+			dbheight = block.EntryBlock.Header.DBHeight
+			c++
+			done := startAmt - len(g.OPRChain.BlocksToBeParsed)
+			if c%30 == 0 || done == startAmt {
+				fLog.WithFields(log.Fields{
+					"dbht": block.EntryBlock.Header.DBHeight,
+				}).Debugf("syncing entries, %.2f%%", float64(done)/float64(startAmt)*100)
+			}
+
+			var err error
+			// Before we try to fetch from the net, we try and fetch from disk
+			oprblock, _ := g.BlockStore.FetchOPRBlock(block.EntryBlock.Header.DBHeight)
+			if oprblock == nil {
+				// Fetch from factomd
+				oprblock, err = g.FetchOPRBlock(block)
+			} else {
+				if oprblock.EmptyOPRBlock {
+					g.OPRChain.BlockParsed(*block)
+					continue // This eblock does not have a valid opr block
+				}
+			}
+
+			// If we have an error from factomd
+			if err != nil {
+				return err
+			}
+
+			if oprblock == nil {
+				err := g.BlockStore.WriteInvalidOPRBlock(block.EntryBlock.Header.DBHeight)
+				if err != nil {
+					return err
+				}
+				g.OPRChain.BlockParsed(*block) // This block is done being processed
+				continue
+			}
+
+			g.oprBlkLock.Lock()
+			// We add the oprs, and the graded blocks. The next iteration of this loop will use these graded oprs.
+			err = g.BlockStore.WriteOPRBlock(oprblock)
+			if err != nil {
+				g.oprBlkLock.Unlock()
+				return err
+			}
+			g.oprBlks = append(g.oprBlks, oprblock)
+			g.oprBlkLock.Unlock()
+
+			// Let's add the winner's rewards. They will be happy that we do this step :)
+			payouts := g.MinRecords(dbheight)
+			for place, winner := range oprblock.GradedOPRs[:payouts] { // The top 25 matter in version 2
+				reward := GetRewardFromPlace(place, g.Network, block.EntryBlock.Header.DBHeight)
+				if reward > 0 {
+					err := g.Balances.AddToBalance(winner.CoinbasePNTAddress, reward)
+					if err != nil {
+						log.WithError(err).Fatal("Failed to update balance")
+					}
+					// Debug logs were here before to print the winners, it was a bit noisy
+				}
+			}
+
+			g.OPRChain.BlockParsed(*block)
+		}
+	}
+	fLog.WithField("height", dbheight).Debugf("synced!")
+
+	return nil
+}
+
+func (g *QuickGrader) FetchOPRBlock(block *EntryBlockMarker) (*OprBlock, error) {
+	min := g.MinRecords(block.EntryBlock.Header.DBHeight)
+	// There is not enough entries in this block, so there is no point in looking at it.
+	if len(block.EntryBlock.EntryList) < min {
+		return nil, nil
+	}
+
+	var oprs []*OraclePriceRecord
+	var err error
+	if len(block.EntryBlock.EntryList) > 50 {
+		// Multithread when there is a lot (like a 6x speedup in my tests against mainnet)
+		oprs, err = g.ParallelFetchOPRsFromEBlock(block)
+	} else {
+		oprs, err = g.FetchOPRsFromEBlock(block)
+	}
```

Comment:
why have two separate implementations for the same code? just have one function that specifies the number of workers and set it to 1 if you want single threaded. is there a reason to not just run the workers every time?